src/xrc/src/api/test.rs

```diff
@@ -209,14 +231,112 @@ fn stablecoin_mock(symbol: &str, rates: &[u64]) -> QueriedExchangeRate {
     )
 }
 
+fn stablecoin_mock_with_failed_exchanges(symbol: &str, rates: &[u64], exchanges: Option<Vec<Exchange>>) -> QueriedExchangeRateWithFailedExchanges {
+    let failed_exchanges = match exchanges {
+        Some(exchanges) => exchanges,
+        None => vec![],
+    };
+    QueriedExchangeRateWithFailedExchanges {
+        queried_exchange_rate: stablecoin_mock(symbol, rates),
+        failed_exchanges,
+    }
+}
+
+/// This function tests that subsequent calls to to an exchange are not made when the first call
+/// fails due to an HTTP error.
+#[test]
+fn get_exchange_rate_skips_exchanges_that_fail_for_cryptocurrency_usdt_rate() {
+    let current_timestamp: u64 = 1678752000;
+    let coinbase = Exchange::Coinbase(Coinbase);
+    let call_exchanges_impl = TestCallExchangesImpl::builder()
+        .with_get_cryptocurrency_usdt_rate_responses(btreemap! {
+            "BTC".to_string() => Ok(btc_queried_exchange_rate_with_failed_exchanges_mock(Some(vec![coinbase.clone()]))),
+            "ICP".to_string() => Ok(icp_queried_exchange_rate_with_failed_exchanges_mock(None)),
+        })
+        .build();
+    let env = TestEnvironment::builder()
+        .with_time_secs(current_timestamp)
+        .with_cycles_available(XRC_REQUEST_CYCLES_COST)
+        .with_accepted_cycles(XRC_MINIMUM_FEE_COST * 500)
+        .build();
+    let request = GetExchangeRateRequest {
+        base_asset: btc_asset(),
+        quote_asset: icp_asset(),
+        timestamp: None,
+    };
+
+    let result = get_exchange_rate_internal(&env, &call_exchanges_impl, &request)
+        .now_or_never()
+        .expect("future should complete");
+
+    let calls = call_exchanges_impl.get_cryptocurrency_usdt_rate_calls.read().unwrap();
+    // Ensure that Coinbase was excluded from the second call.
+    assert!(!calls[1].0.contains(&coinbase));
```

Comment:
How about decomposing `calls` so that it is more obvious what this `assert` statement shows?